[FINAL] feat: wasm_memory_persistence option for canister upgrades (#281)
dfinity/interface-spec: spec/index.md

When a canister is upgraded to a new WebAssembly module, the IC:

1.  Invokes `canister_pre_upgrade` (if exported by the current canister code and `skip_pre_upgrade` is not `opt true`) on the old instance, to give the canister a chance to clean up (e.g. move data to [stable memory](#system-api-stable-memory)).

2.  Instantiates the new module, including the execution of `(start)`, with a fresh WebAssembly state.

3.  Invokes `canister_post_upgrade` (if present) on the new instance, passing the `arg` provided in the `install_code` call ([IC method](#ic-install_code)).

> The stable memory is preserved throughout the process; any other WebAssembly state is discarded unless `wasm_memory_persistence` is `opt keep` in which case only WebAssembly data segments are discarded.

Reviewer: A detail aspect: The active data segments of the initial program version would still be accessible as the entire main memory is retained. For the new program version, the active data segments are not loaded. This is why passive data segments are currently used for the orthogonal persistence.
Maybe one could phrase it as: "in which case the main Wasm memory is additionally preserved.". 
Maybe with additional information: "If `wasm_memory_peristence is `opt keep`, active data segments of the new program version are not loaded."

#### IC Management Canister: Code upgrade

Only the controllers of the given canister can install new code. This changes the code of an *existing* canister, preserving the state in the stable memory. This involves invoking the `canister_pre_upgrade` method, if the `skip_pre_upgrade` flag is not set to `opt true`, on the old and `canister_post_upgrade` method on the new canister, which must succeed and must not invoke other methods. If the `wasm_memory_persistence` flag is set to `opt keep`, then the WebAssembly memory is preserved.

In the following, the `initial_wasm_store` is the store of the WebAssembly module after instantiation (as per WebAssembly spec) of the WebAssembly module contained in `A.wasm_module`, before executing a potential `(start)` function. The store `initialize_store(State, A.wasm_module)` is the store of the WebAssembly module after instantiation (as per WebAssembly spec) of the WebAssembly module contained in `A.wasm_module` while reusing the WebAssembly memory of `State`.

> If the canister exports a private custom section with the name "enhanced-orthogonal-persistence", then the `wasm_memory_persistence` option must be set to `opt keep`.

Reviewer 2: I first misunderstood the mechanism as looking at the custom section of the *new* module. Re-reading this sentence, it is already unambiguous, but maybe we want to still emphasize a bit more explicitly that the custom section is in the *already running* wasm module.

#### IC Management Canister: Code upgrade

> Only the controllers of the given canister can install new code. This changes the code of an *existing* canister, preserving the state in the stable memory. This involves invoking the `canister_pre_upgrade` method, if the `skip_pre_upgrade` flag is not set to `opt true`, on the old and `canister_post_upgrade` method on the new canister, which must succeed and must not invoke other methods. If the `wasm_memory_persistence` flag is set to `opt keep`, then the WebAssembly state is preserved except for WebAssembly data segments.

Reviewer: Maybe: 
```suggestion
Only the controllers of the given canister can install new code. This changes the code of an *existing* canister, preserving the state in the stable memory. This involves invoking the `canister_pre_upgrade` method, if the `skip_pre_upgrade` flag is not set to `opt true`, on the old and `canister_post_upgrade` method on the new canister, which must succeed and must not invoke other methods. If the `wasm_memory_persistence` flag is set to `opt keep`, then the WebAssembly memory state is preserved and data segments of the new WebAssembly code need to be explicitly loaded by using passive data segments.
```

Author: In that case, I'd go ahead and simplify the spec to not talk about data segments at all. Does that sound good?

> In the following, the `initial_wasm_store` is the store of the WebAssembly module after *instantiation* (as per WebAssembly spec) of the WebAssembly module contained in the [canister module](#canister-module-format), before executing a potential `(start)` function. The function `load_data_segments(S, wasm_module)` returns the WebAssembly store `S` after loading the data segments from the given WebAssembly module.

Reviewer: We could probably skip the `load_data_segments()`, please cf. below.

}

(

>   (A.mode = upgrade U and U.skip_pre_upgrade ≠ true)

Author: > Are there some Candid options missing here?

What do you mean by "missing"? In this condition, only skip_pre_uprade is relevant and that's why the other options do not appear explicitly (they aren't read from the record U).

  Env1 = Env with {
    global_timer = S.global_timer[A.canister_id];
    canister_version = S.canister_version[A.canister_id];
  }
  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {new_state = Intermediate_state; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
)
or
(
  (A.mode = upgrade U and U.skip_pre_upgrade = true)
  Intermediate_state = Old_state
  New_certified_data = NoCertifiedData
  Cycles_used = 0
)

(
  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
  Persisted_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Intermediate_state.stable_memory}
)
or
(
  (A.mode = upgrade U and U.wasm_memory_persistence = keep)

>   Persisted_state = Intermediate_state

Author: ```suggestion
  Persisted_state = initialize_stote(Intermediate_state, A.wasm_module)
```
where `initialize_store(State, Wasm_module)` is the store of the WebAssembly module after *instantiation* (as per WebAssembly spec) of the WebAssembly module `Wasm_module` reusing the WebAssembly memory of `State`.

Author: Done without commenting on active data segments. @crusso Could you please comment on whether special treatment of active data segments is needed after all?

Reviewer 3: @luc-blaeser and I discussed this offline and perhaps the simplest to specify is to disallow active data segments for EOP binaries - i.e, when keep is true. The motoko compiler already only emits passive data segments, rewriting the active data segment of the RTS to a passive one off-line.

Once could relax the restriction of no active data segments later  on in the spec and implementation.

Author: > disallow active data segments for (replacement) EOP (only) binaries - i.e, when keep is true

Why would it be simpler to disallow them rather than not commenting on them at all (meaning if the new module provides active data segments, then they'll overwrite parts of the persisted Wasm memory)?

Reviewer: I had a corresponding prototype implementation, but there was a concern on complexity (and performance) of the upgrade that we have to selectively patch the active data segments. I believe a reason is that for the initial memory setup, the data segments can be applied in a simple page-based granularity (`Segments::as_pages()` extending the data segments to pages with zero initialization of the unused parts), while for the upgrade, the patching would need to be happen on the existing pages in the more fine-granular byte-wise approach.

The main reason was however that it is not very useful because the runtime system would need to reserve a fix section for active data segments in order not accidentally overwrite the heap. Choosing the size of the reserved space would be very difficult, it is often too wasteful or too limited. This is why we only use passive data segments, as it gives us the flexibility to choose the addresses dynamically in a simple way.

}

(

>   (A.mode = upgrade U and U.skip_pre_upgrade ≠ true)

Reviewer 3: What I mean is, assuming A is the textual candid argument, isn't the correct condition (taking account of the defaults) something like
```suggestion
  (A.mode = variant { upgrade = 0 } and 0 = null or (0 = opt U and  U.skip_pre_upgrade ≠ opt true))
```
But that is really ugly...

Reviewer 3: Or is there some convention that let's you finesse this?

Author: I assume an informal dot notation on `A` since `A` is turned into a binary payload using the function `candid(A)`.

  Env1 = Env with {
    global_timer = S.global_timer[A.canister_id];
    canister_version = S.canister_version[A.canister_id];
  }
  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {new_state = Intermediate_state; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
)
or
(

>   (A.mode = upgrade U and U.skip_pre_upgrade = true)

Reviewer 3: ```suggestion
  (A.mode = variant { upgrade = opt U } and U.skip_pre_upgrade = opt true)
```
Perhaps?

  Intermediate_state = Old_state
  New_certified_data = NoCertifiedData
  Cycles_used = 0
)

(

>   (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)

Reviewer 3: ```suggestion
  (A.mode = variant { upgrade = O } and O = null or O = opt U and U.wasm_memory_persistence ≠ opt variant { keep })
```
Yuck

  Persisted_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Intermediate_state.stable_memory}
)
or
(

>   (A.mode = upgrade U and U.wasm_memory_persistence = keep)

Reviewer 3: ```suggestion
  (A.mode = variant { upgrade = opt U } and U.wasm_memory_persistence = opt keep)
```

  Persisted_state = initialize_store(Intermediate_state, A.wasm_module)
)

> (A.mode = upgrade U and U.wasm_memory_persistence = keep)

Reviewer 3: ```suggestion
(A.mode = variant { upgrade = opt U } and U.wasm_memory_persistence = opt variant { keep } )
```

}

(
  (A.mode = upgrade U and U.skip_pre_upgrade ≠ true)
  Env1 = Env with {
    global_timer = S.global_timer[A.canister_id];
    canister_version = S.canister_version[A.canister_id];
  }
  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {new_state = Intermediate_state; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
)
or
(
  (A.mode = upgrade U and U.skip_pre_upgrade = true)
  Intermediate_state = Old_state
  New_certified_data = NoCertifiedData
  Cycles_used = 0
)

(
  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
  Persisted_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Intermediate_state.stable_memory}
)
or
(
  (A.mode = upgrade U and U.wasm_memory_persistence = keep)
  Persisted_state = initialize_store(Intermediate_state, A.wasm_module)
)

(A.mode = upgrade U and U.wasm_memory_persistence = keep)
or
(A.mode = upgrade U and U.wasm_memory_persistence = replace)
or
(S.canisters[A.canister_id].private_custom_sections["enhanced-orthogonal-persistence"] = null)

> not (A.mode = upgrade U and U.wasm_memory_persistence = keep and Private_custom_sections["enhanced-orthogonal-persistence"] = null)

Author: @luc-blaeser Should we add an analogous check that the old module must declare the custom section "enhanced-orthogonal-persistence" if we use the "keep" option?

Reviewer 3: I just discussed this with @luc-blaeser and I think his implementation implements that check, so the spec is probably missing it. But best to get confirmation from @luc-blaeser himself...

Reviewer: Yes, I agree, it would make sense to check that both at least the **new** version has the EOP custom section, when `Keep` is specified. Otherwise, a classical version would misinterpret the non-zero main memory and not see any serialized state in stable memory. This is currently implemented in the MR (https://gitlab.com/dfinity-lab/public/ic/-/merge_requests/17893)

We could also additionally require that the **old** version carries the EOP custom section, when `Keep` is specified. However, this is less critical, as the EOP runtime should detect this case and simply deserialize state from stable memory and reset the main memory state. This check is currently not implemented in the MR.

Author: >  This is currently implemented in the MR

and also in the spec at the line

```
not (A.mode = upgrade U and U.wasm_memory_persistence = keep and Private_custom_sections["enhanced-orthogonal-persistence"] = null)
```

> This check is currently not implemented in the MR and in my opinion, not necessarily needed.

This is what I was asking.

>  as the EOP runtime should detect this case and simply deserialize state from stable memory and reset the main memory state

Then it sounds safe, indeed.

)
or
(
  (A.mode = upgrade U and U.skip_pre_upgrade = true)
  Stable_memory = Old_State.stable_mem
  New_certified_data = NoCertifiedData
  Cycles_used = 0
)

(
  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
  Intermediate_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Stable_memory}
)
or
(
  (A.mode = upgrade U and U.wasm_memory_persistence = keep)

>   Intermediate_state = {store = load_data_segments(Old_state.store, A.wasm_module); self_id = A.canister_id; stable_mem = Stable_memory}

Reviewer: For simplicity, we thought not to load the new active data segments and not to patch the memory (as this would need a reserved space and is also more complicated for the IC execution layer). Instead, we can use passive data segments that can be lazily loaded at a dynamic address.

It would be more like:
```suggestion
  Intermediate_state = {store = Old_state.store; self_id = A.canister_id; stable_mem = Stable_memory}
```

>     If the WebAssembly module does not export a function called under the name `canister_pre_upgrade`, then it simply returns the stable memory:

Reviewer 3: ```suggestion
    If the WebAssembly module does not export a function called under the name `canister_pre_upgrade`, then it simply returns the current state:
```
